x-pack/legacy/plugins/lens/public/indexpattern_plugin/datapanel.tsx

```diff
@@ -163,18 +225,96 @@ export const InnerIndexPatternDataPanel = function InnerIndexPatternDataPanel({
   }
 
   const allFields = indexPatterns[currentIndexPatternId].fields;
-  const filteredFields = allFields
+
+  const availableFieldTypes = uniq(allFields.map(({ type }) => type)).filter(
+    type => type in fieldTypeNames
+  );
+
+  const displayedFields = allFields
+    .filter(field => {
+      if (!showEmptyFields) {
+        const indexField =
+          indexPatterns[currentIndexPatternId] &&
+          indexPatterns[currentIndexPatternId].hasExistence &&
+          indexPatterns[currentIndexPatternId].fields.find(f => f.name === field.name);
+        if (localState.typeFilter.length > 0) {
+          return (
+            indexField &&
+            indexField.exists &&
+            localState.typeFilter.includes(field.type as DataType)
+          );
+        }
+        return indexField && indexField.exists;
+      }
+      if (localState.typeFilter.length > 0) {
+        return localState.typeFilter.includes(field.type as DataType);
+      }
+      return true;
+    })
     .filter(
       (field: IndexPatternField) =>
-        field.name.toLowerCase().includes(state.nameFilter.toLowerCase()) &&
+        field.name.toLowerCase().includes(localState.nameFilter.toLowerCase()) &&
         supportedFieldTypes.includes(field.type)
-    )
-    .slice(0, pageSize);
+    );
 
-  const availableFieldTypes = uniq(filteredFields.map(({ type }) => type));
-  const availableFilteredTypes = state.typeFilter.filter(type =>
-    availableFieldTypes.includes(type)
-  );
+  const paginatedFields = displayedFields.sort(sortFields).slice(0, pageSize);
+
+  useEffect(() => {
+    if (
+      localState.isLoading ||
+      indexPatterns[currentIndexPatternId].hasExistence ||
+      !updateFieldsWithCounts
+    ) {
+      return;
+    }
+
+    setLocalState(s => ({ ...s, isLoading: true }));
+
+    npStart.core.http
+      .post(`/api/lens/index_stats/${indexPatterns[currentIndexPatternId].title}`, {
```

Comment:
Discussed offline. I'm concerned about the size of the body here, and given how large index pattern saved objects are, it's very likely that this would be rejected by browsers.